src/pow.rs

```diff
@@ -0,0 +1,1540 @@
+            bn.0.low_u32()
+        };
+
+        if (compact & 0x0080_0000) != 0 {
+            compact >>= 8;
+            size += 1;
+        }
+
+        compact | (size << 24) as u32
+    }
+
+    /// Returns true if block hash is less than or equal to this [`Target`].
+    ///
+    /// Proof-of-work validity for a block requires the hash of the block to be less than or equal
+    /// to the target.
+    pub fn is_met_by(&self, hash: BlockHash) -> bool {
+        use crate::hashes::Hash;
+        let hash = Target(Uint256::from_le_bytes(hash.into_inner()));
+        hash <= *self
+    }
+
+    /// Converts this [`Target`] to [`Work`].
+    ///
+    /// "Work" is defined as the work done to mine a block with this target value (recorded in the
+    /// block header in compact form as nBits). This is not the same as the difficulty to mine a
+    /// block with this target (see `Self::difficulty`).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Target;
+    /// let t = Target::from(0xdeadbeef_u32);
+    /// assert_eq!(t.to_work().to_target(), t)
+    /// ```
+    pub fn to_work(self) -> Work { Work(self.0.inverse()) }
+
+    /// Computes the popular "difficulty" measure for mining.
+    ///
+    /// Difficulty represents how difficult the current target makes it to find a block, relative to
+    /// how difficult it would be at the highest possible target (highest target == lowest difficulty).
+    ///
+    /// For example, a difficulty of 6,695,826 means that at a given hash rate, it will, on average,
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    ///
+    /// # Note
+    ///
+    /// Difficulty is calculated using the following algorithm `max / current` where [max] is
+    /// defined for the Bitcoin network and `current` is the current [target] for this block. As
+    /// such, a low target implies a high difficulty. Since [`Target`] is represented as a 256 bit
+    /// integer but `difficulty()` returns only 128 bits this means for targets below approximately
+    /// `0xffff_ffff_ffff_ffff_ffff_ffff` `difficulty()` will saturate at `u128::MAX`.
+    ///
+    /// [max]: crate::blockdata::constants::max_target
+    /// [target]: crate::blockdata::block::BlockHeader::target
+    pub fn difficulty(&self) -> u128 {
+        const MAX: Uint256 = Uint256([0, u128::max_value()]);
+
+        let d = constants::max_target().0 / self.0;
+
+        if d > MAX {
+            u128::max_value()
+        } else {
+            d.low_u128()
+        }
+    }
+}
+do_impl!(Target);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Big-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+struct Uint256(pub [u128; 2]);
```

Comment:
For consideration: use `struct U256(u128, u128);` instead:

* naming consistent with other types (minus casing)
* now that there are only two items we don't need for loops that much and seems even simpler without them
* We avoid awkward `self.0[0]` and `self.0[1]` - just have `self.0` and `self.1`

Reply:
I like this idea, will play with it.